datafusion/expr/src/logical_plan/invariants.rs

```diff
@@ -28,6 +30,24 @@ use crate::{
     Aggregate, Expr, Filter, Join, JoinType, LogicalPlan, Window,
 };
 
+use super::Extension;
+
+pub type InvariantFn = Arc<dyn Fn(&LogicalPlan) -> Result<()> + Send + Sync>;
+
+#[derive(Clone)]
+pub struct Invariant {
+    pub kind: InvariantLevel,
+    pub fun: InvariantFn,
+}
+
+impl Invariant {
+    /// Return an error if invariant does not hold true.
+    pub fn check(&self, plan: &LogicalPlan) -> Result<()> {
+        (self.fun)(plan)
+    }
+}
```

Comment:
This is a user-defined invariant.

Reply:
Note: a followup PR will enable a user-defined LP invariant to be run against all LP nodes (not just extension nodes).